File: packages/core/src/terminal/idle-detector.ts
// eslint-disable-next-line no-control-regex
const ANSI = /\x1b\[[0-9;?]*[A-Za-z]/g;

/** How many trailing characters of the buffer to inspect. */
const TAIL = 2000;

/**
 * Best-effort idle/turn-complete detector.
 *
 * Real TUI output (from spike 2026-05-22): after ANSI-stripping and whitespace
 * collapsing, the ready input box appears as "❯Try\"create a util...\"" or a
 * lone "❯". Working state shows spinner glyphs (✻✽✶✳✢·) but no input-box caret.
 * Permission menus show "❯1.Yes" / "❯2.…" — digit immediately after ❯.
 *
 * Heuristics (all must hold):
 *   1. Tail contains ❯ NOT immediately followed by a digit (input box, not a menu).
 *   2. Tail does NOT contain "doyouwantto" (not a permission prompt).
 *
 * NOTE: the post-turn idle state (after a task completes) was NOT captured in the
 * spike — both runs stalled at the permission prompt. The pattern here is based on
 * the pre-prompt ready state (same ❯ input box). Confirm at E2E.
 */
export function detectIdle(buffer: string): boolean {
  if (!buffer) return false;

  const flat = buffer.slice(-TAIL).replace(ANSI, "").replace(/\s+/g, "");

  // Must have input-box caret not followed by a digit
  if (!/❯(?![0-9])/.test(flat)) return false;

  // Must not be a permission prompt
  if (/doyouwantto/i.test(flat)) return false;

  return true;
}

File: rust/crates/core/src/terminal/idle_detector.rs
use regex::Regex;
use std::sync::OnceLock;

/// Trailing characters of the buffer to inspect.
const TAIL: usize = 2000;

/// Strips ANSI escape codes from a string slice.
fn strip_ansi(s: &str) -> String {
    static ANSI_RE: OnceLock<Regex> = OnceLock::new();
    let re = ANSI_RE.get_or_init(|| Regex::new(r"\x1b\[[0-9;?]*[A-Za-z]").unwrap());
    re.replace_all(s, "").into_owned()
}

/// Return true if `flat` contains `❯` (U+276F) NOT immediately followed by an ASCII digit.
///
/// The Rust `regex` crate does not support lookahead, so we scan manually.
fn has_input_box_caret(flat: &str) -> bool {
    const CARET: char = '\u{276f}'; // ❯
    let mut chars = flat.char_indices().peekable();
    while let Some((_, c)) = chars.next() {
        if c == CARET {
            match chars.peek() {
                // ❯ at end of string — counts as input box
                None => return true,
                // ❯ not followed by a digit — counts as input box
                Some((_, next)) if !next.is_ascii_digit() => return true,
                // ❯ followed by a digit — menu item, keep scanning
                _ => {}
            }
        }
    }
    false
}

/// Best-effort idle / turn-complete detector.
///
/// Mirrors TypeScript `detectIdle`:
///   1. Tail must contain `❯` NOT immediately followed by a digit (input box, not a menu).
///   2. Tail must NOT contain "doyouwantto" (not a permission prompt).
pub fn detect_idle(buffer: &str) -> bool {
    static PERMISSION_RE: OnceLock<Regex> = OnceLock::new();

    if buffer.is_empty() {
        return false;
    }

    let tail = if buffer.len() > TAIL {
        &buffer[buffer.len() - TAIL..]
    } else {
        buffer
    };

    let stripped = strip_ansi(tail);
    // Collapse all whitespace
    let flat: String = stripped.chars().filter(|c| !c.is_whitespace()).collect();

    if !has_input_box_caret(&flat) {
        return false;
    }

    let permission_re =
        PERMISSION_RE.get_or_init(|| Regex::new(r"(?i)doyouwantto").unwrap());
    if permission_re.is_match(&flat) {
        return false;
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn idle_detects_input_box_caret() {
        // ❯ not followed by a digit = input box = idle
        let buf = "\u{276f}Try \"create a util...\"";
        assert!(detect_idle(buf));
    }

    #[test]
    fn idle_not_triggered_by_permission_menu() {
        let buf = "Do you want to create hello.txt? \u{276f}1. Yes 2. No";
        assert!(!detect_idle(buf));
    }

    #[test]
    fn idle_not_triggered_on_empty() {
        assert!(!detect_idle(""));
    }
}
